vendor/github.com/xtaci/smux/session.go

```diff
@@ -281,73 +276,76 @@ func (s *Session) recvLoop() {
 	}
 }
 
-func (s *Session) keepalive() {
-	tickerPing := time.NewTicker(s.config.KeepAliveInterval)
-	tickerTimeout := time.NewTicker(s.config.KeepAliveTimeout)
-	defer tickerPing.Stop()
-	defer tickerTimeout.Stop()
+// keepAliveSend will periodically send a keepalive meesage to the remote peer.
+func (s *Session) keepAliveSend() {
+	keepAliveTimeout := time.After(s.config.KeepAliveInterval)
 	for {
 		select {
-		case <-tickerPing.C:
-			s.writeFrame(newFrame(cmdNOP, 0))
-			s.notifyBucket() // force a signal to the recvLoop
-		case <-tickerTimeout.C:
-			if !atomic.CompareAndSwapInt32(&s.dataReady, 1, 0) {
-				s.Close()
-				return
-			}
 		case <-s.die:
 			return
+		case <-keepAliveTimeout:
+			s.writeFrame(newFrame(cmdNOP, 0), time.Now().Add(s.config.WriteTimeout))
+			s.notifyBucket() // force a signal to the recvLoop
+			keepAliveTimeout = time.After(s.config.KeepAliveInterval)
 		}
 	}
 }
 
-func (s *Session) sendLoop() {
-	buf := make([]byte, (1<<16)+headerSize)
+// keepAliveTimeout will periodically check that some sort of message has been
+// sent by the remote peer, closing the session if not.
+func (s *Session) keepAliveTimeout() {
+	timeoutChan := time.After(s.config.KeepAliveTimeout)
 	for {
 		select {
 		case <-s.die:
 			return
-		case request, ok := <-s.writes:
-			if !ok {
-				continue
-			}
-			buf[0] = request.frame.ver
-			buf[1] = request.frame.cmd
-			binary.LittleEndian.PutUint16(buf[2:], uint16(len(request.frame.data)))
-			binary.LittleEndian.PutUint32(buf[4:], request.frame.sid)
-			copy(buf[headerSize:], request.frame.data)
-			n, err := s.conn.Write(buf[:headerSize+len(request.frame.data)])
-
-			n -= headerSize
-			if n < 0 {
-				n = 0
-			}
-
-			result := writeResult{
-				n:   n,
-				err: err,
```

Comment:
This red diff is from the function `sendLoop` that I deleted, it's not from `keepAliveTimeout`